main.cc

```diff
@@ -0,0 +1,108 @@
+#include "Indent.h"
+#include <iostream>
+#include <map>
+
+using StrMap=std::map<std::string, std::string>;
+
+/* desired output:
+---
+level0.1:
+  level1.1:
+    a: 1
+    b: 2
+  level1.2:
+    a: 4
+    b: 5
+level0.2:
+  - k1: v1
+    k2: v2
+  - k1: v1
+    k2: v2
+ */
+
+void
+subItemMap(std::ostream &o, const std::string &prefix,
+           const StrMap &m)
+{
+    for (const auto &it: m) {
+        o << prefix << it.first << ": " << it.second << "\n";
+    }
+}
+void
+subItemMap(std::ostream &o, const Indent &prefix,
+           const StrMap &m, bool isVector=false)
+{
+    for (const auto &it : m) {
+        if (isVector)
+            o << prefix.elem();
+        else
+            o << prefix;
+        o << it.first << ": " << it.second << "\n";
+        isVector = false;
+    }
+}
+
+void
+subVectorMap(std::ostream &o, const Indent &prefix, const std::vector<StrMap> &m)
+{
+    for (const auto &it : m) {
+        subItemMap(o, prefix, it, true);
+    }
+}
+
+int main()
+{
+    std::ostream& o=std::cout;
+
+    o << "**** using spaces()\n"
+        << "level0.1:\n"
+        << spaces(2) << "level1.1:\n"
+        << spaces(4) << "a: 1\n"
+        << spaces(4) << "b: 2\n"
+        << spaces(2) << "level1.2:\n"
+        << spaces(4) << "a: 4\n"
+        << spaces(4) << "b: 5\n"
+        << "level0.2:\n"
+        << spaces(0) << "- k1: v1\n"
+        << spaces(2) << "k2: v2\n"
+        << spaces(0) << "- k1: v1\n"
+        << spaces(2) << "k2: v2\n";
+
+    o << "**** using Indent()\n";
```

Comment:
Copying @rousskov's comment:

Right now, all three usage examples do not look good IMO:

A "using spaces" example (above) lacks any structure (despite the underlying data having one).

A "using Indent" example (below) seems to be effectively the same as the first, but with iN names for various indentation levels added instead of explicit spaces(n) calls. Renaming "i1" to "spaces2", etc. would highlight these similarities. The subItemMap() difference appears to be a red herring -- the spaces() example could use a function like that as well.

A "using Indent() and scopes" example (further below) implies that it uses "scopes", but removing scopes (and de-duplicating identical i1 declarations to get the code compile) would change nothing. Thus, the underlying idea of using scopes is not really illustrated/shown AFAICT.

Overall, these examples is not the kind of code I would want to write, review, or maintain! However, it is possible that some of these examples can be written differently (while using the same or similar indentation APIs) to reflect data structure and reduce risks. I am not sure about that, but I am sure that we should start with a good example and then decide whether we can support it. If I find the time, I will attempt to write an example that has fewer problems.

Reply:
spaces() is where we left off in Squid. It's a way to have explicit indentation and let the caller manage it.

The Indent object is meant to be used as a stack. The idea behind it is that the user can increase the indentation level by instantiating a new object, but the old one needs to be kept around to allow to return to the previous indentation level. It's meant to be used as a kind-of RAII. It can be passed around to functions to satisfy the use case you expressed there.

The obvious alternative would be to have an object that supports `in()` and `out()` methods (or somesuch) that increase/decrease indentation. The main problem I see with that is that if it's passed around to a function and that function contains a bug by which it doesn't reset the right indentation level, then the whole YAML will be broken from that invocation point onwards

Reply:
At this time, I am focusing on report-building code quality (as expressed by our examples). Your response is mostly discussing something else (let's call that something else "X"). I still hope for a different outcome than "report-building code will always be ugly and unsafe because of X", but I am prepared to be proven wrong. If nobody (including me) can propose much better examples, then I will be proven wrong.

Reply:
> I will attempt to write an example that has fewer problems.

Here is my first sketch. I tried to build a representative sample of primary reporting variations in existing Squid reporting code. This sketch does not cover non-atomic actions (an important but mostly orthogonal TODO).

```C++
// Version: 10
#if DESIRED_OUTPUT

---
kid: 1
requests received: 2
read size histogram:
    0: { count: 0, portion: 0%, cumulative: 0% }
    1: { count: 123, portion: 3%, cumulative: 3% }
    <...>
    99: { count: 1, portion: 0%, cumulative: 100% }
sample start time = <...>
sample end time = <...>
<...>
peers:
    - example.com:3128:
        digest state: needed
        <...>
    - example.net:80:
        digest state: usable
        <...>
    - no-digest.test:443:
bytes received: 200000
...
#endif

void
LegacyCodeToBeConverted(StoreEntry * const e)
{
    storeAppendPrintf(e, "requests received: %.0f\n", stats.requests);

    storeAppendPrintf(e, "read size histogram:\n");
    for (i = 0; i < 100; ++i)
        storeAppendPrintf(e, "%5d: %9.0f %2.0f%%\n", i, hist[i], percent(hist[i], stats.reads));

    SampleStats sample;
    GetSampleStats(sample, 5);
    ReportSample(e, sample);

    for (const auto &peer: CurrentPeers()) {
        if (peer->digest)
            peerDigestStatsReport(peer->digest, e);
        else
            storeAppendPrintf(e, "No digest from %s\n", peer->name);
    }

    storeAppendPrintf(e, "bytes received: %.0f\n", stats.bytes_in);
}

// Three primary YAML nodes:
// * mapping (i.e. a possibly empty list of key:value nodes, where key and value are also nodes)
// * sequence (i.e. a possibly empty list of nodes)
// * scalar (nearly everything else; e.g., an integer or a quoted string)
//
// TODO: Non-atomic actions need to _resume_ producing output for the current
// YAML sequence. That implies we need to be able to store the current YAML
// stream state. That might imply it has to be passed as unique_ptr, but I hope
// we can hide that complication via Sequence::suspend/resume() and still pass
// Yaml::Stream references.
void
ConvertedCode(Yaml::Mapping &yaml)
{
    // We could also use an explicit Yaml::kv() call here, but we do not have to
    // because Yaml::Mapping knows that the first node it receives is the key
    // and second node it receives is the value for that key (and so on).
    yaml << "requests received" << stats.requests;

    if (stats.reads) {
        const auto histogram = yaml.openMap("read size histogram");
        for (i = 0; i < 100; ++i)
            histogram << i << Yaml::HistogramEntryValue(hist[i], stats.reads);
    }
 
    SampleStats sample;
    GetSampleStats(sample, 5);
    ReportSample(yaml, sample);

    const auto peers = yaml.openSequence("peers");
    for (const auto &peer: CurrentPeers()) {
        const auto peerMap = peers.openMap(peer->name);
        if (peer->digest)
            peer->digest->report(peers);
        // else: empty map for this no-digest peer
    }
    // TODO: We can automate this closure. Should we?
    yaml.close(peers) // or peers.close();

    yaml << "bytes received" << stats.bytes_in;
}
```

Reply:
> > I will attempt to write an example that has fewer problems.
> 
> Here is my first sketch. I tried to build a representative sample of primary reporting variations in existing Squid reporting code. 

> This sketch does not cover non-atomic actions (an important but mostly orthogonal TODO).
I agree.

The output you sketch here is legitimate YAML, but it is not good YAML.
In particular the `peers` key is unnecessarily complex

```
peers:
    - example.com:3128:
        digest state: needed
        <...>
    - example.net:80:
        digest state: usable
        <...>
    - no-digest.test:443:
```
`peers` is a map entry containing a sequence of maps , each with exactly one key having as value a map

It could be more effectively rendered as:
```
peers:
  example.com:3128:
    digest state: needed
    <...>
  example.net:80:
    digest state: usable
    <...>
  no-digest.test:443:
    <...>
```
a map entry containing a map of maps

but I suspect that given the context your intent is to tickle code for a sequence, in which case the output should be:

```
peers:
  - host: example.com:3128
    digest state: needed
  - host: example.net:80
    digest state: usable
  - host: no-digest.test:443
```
aka a map element containing a sequence of maps

> ```c++
> // Version: 10
> #if DESIRED_OUTPUT
> 
> ---
> kid: 1
> requests received: 2
> read size histogram:
>     0: { count: 0, portion: 0%, cumulative: 0% }
>     1: { count: 123, portion: 3%, cumulative: 3% }
>     <...>
>     99: { count: 1, portion: 0%, cumulative: 100% }
> sample start time = <...>
> sample end time = <...>
> <...>
> peers:
>     - example.com:3128:
>         digest state: needed
>         <...>
>     - example.net:80:
>         digest state: usable
>         <...>
>     - no-digest.test:443:
> bytes received: 200000
> ...
> #endif
> 
> void
> LegacyCodeToBeConverted(StoreEntry * const e)
> {
>     storeAppendPrintf(e, "requests received: %.0f\n", stats.requests);
> 
>     storeAppendPrintf(e, "read size histogram:\n");
>     for (i = 0; i < 100; ++i)
>         storeAppendPrintf(e, "%5d: %9.0f %2.0f%%\n", i, hist[i], percent(hist[i], stats.reads));
> 
>     SampleStats sample;
>     GetSampleStats(sample, 5);
>     ReportSample(e, sample);
> 
>     for (const auto &peer: CurrentPeers()) {
>         if (peer->digest)
>             peerDigestStatsReport(peer->digest, e);
>         else
>             storeAppendPrintf(e, "No digest from %s\n", peer->name);
>     }
> 
>     storeAppendPrintf(e, "bytes received: %.0f\n", stats.bytes_in);
> }
> 
> // Three primary YAML nodes:
> // * mapping (i.e. a possibly empty list of key:value nodes, where key and value are also nodes)
> // * sequence (i.e. a possibly empty list of nodes)
> // * scalar (nearly everything else; e.g., an integer or a quoted string)
> //
> // TODO: Non-atomic actions need to _resume_ producing output for the current
> // YAML sequence. That implies we need to be able to store the current YAML
> // stream state. That might imply it has to be passed as unique_ptr, but I hope
> // we can hide that complication via Sequence::suspend/resume() and still pass
> // Yaml::Stream references.

Could we simplify by demanding that non-atomic actions do not leave the stream in a non-resumable state? aka they should be interrupting their output at a resumable boundary (e.g. at the end of describing an entry in the cache index)

> void
> ConvertedCode(Yaml::Mapping &yaml)
> {
>     // We could also use an explicit Yaml::kv() call here, but we do not have to
>     // because Yaml::Mapping knows that the first node it receives is the key
>     // and second node it receives is the value for that key (and so on).
>     yaml << "requests received" << stats.requests;

Is Yaml::Mapping derived from std::ostream? It must be, otherwise we could not rely on all the operator<<'s.
How could we intercept operator<< calls to it?
How can we distinguish the first call from the second, third, and so on? The number of calls per entry is not predetermined.
We might have an Yaml::Mapping::operator() returning an ostream, so the syntax would end up being

```c++
yaml() << key << value1 << value2...;
```

>     if (stats.reads) {
>         const auto histogram = yaml.openMap("read size histogram");
>         for (i = 0; i < 100; ++i)
>             histogram << i << Yaml::HistogramEntryValue(hist[i], stats.reads);
>     }

ok, so you're preserving the RAII foundation of my proposal, ok.

>     SampleStats sample;
>     GetSampleStats(sample, 5);
>     ReportSample(yaml, sample);
> 
>     const auto peers = yaml.openSequence("peers");
>     for (const auto &peer: CurrentPeers()) {
>         const auto peerMap = peers.openMap(peer->name);
>         if (peer->digest)
>             peer->digest->report(peers);
>         // else: empty map for this no-digest peer
>     }
>     // TODO: We can automate this closure. Should we?

We should. Why would we give up on RAII safety for the whole design to avoid opening one more c++ scope?
(interestingly, [yaml-cpp](https://github.com/jbeder/yaml-cpp/wiki/How-To-Emit-YAML) uses manipulators, and has no RAII safety)

>     yaml.close(peers) // or peers.close();


> 
>     yaml << "bytes received" << stats.bytes_in;
> }
> ```

This example is not fitting, because it specializes Sequence, forcing it to Sequence-of-maps.
(and sequence is the real sticker in the YAML syntax definition).

It could maybe evolve into:
```c++
    const auto peers = yaml.openSequence("peers");
    for (const auto &peer: CurrentPeers()) {
       auto peerEntry = peers.entry().openMap();
       peerEntry << "name" << peer->name;
        if (peer->digest)
            peer->digest->report(peers);
        // else: empty map for this no-digest peer
    }
    // TODO: We can automate this closure. Should we?
```

.. It's starting to feel overcomplicated.

What if we just used [yaml-cpp](https://github.com/jbeder/yaml-cpp)?
Maybe it could be as simple as wrapping two of their manipulators to offer RAII safety

Reply:
> `peers` is a map entry containing a sequence of maps , each with exactly one key having as value a map
> It could be more effectively rendered as a map entry containing a map of maps

We should not use a map for ordered collections. cache_peers is an example of such a collection -- the order of cache_peer lines matters and should be reflected in report output. This issue feels completely irrelevant to YAML formatting APIs though; you are right to just assume that we do need a sequence here:

> but I suspect that given the context your intent is to tickle code for a sequence, in which case the output should be:
 
> ```yaml
> peers:
>   - host: example.com:3128
>     digest state: needed
>   - host: example.net:80
>     digest state: usable
>   - host: no-digest.test:443
> ```
> 
> aka a map element containing a sequence of maps

Sure, let's assume that the above is the output. I am not skilled in manually generating YAML output. Thank you for correcting that manual rendering mistake. AFAICT, expected output correction has no affect on the corresponding API.

> > ```c++
> > // TODO: Non-atomic actions need to _resume_ producing output for the current
> > // YAML sequence. That implies we need to be able to store the current YAML
> > // stream state. That might imply it has to be passed as unique_ptr, but I hope
> > // we can hide that complication via Sequence::suspend/resume() and still pass
> > // Yaml::Stream references.
> > ```
> 
> Could we simplify by demanding that non-atomic actions do not leave the stream in a non-resumable state? aka they should be interrupting their output at a resumable boundary (e.g. at the end of describing an entry in the cache index)

Define "resumable boundary".

Right now, I see no advantage in introducing a "resumable boundary" restriction because we would still need to save state -- there is no way around that. Right now, virtually any YAML stream state is "resumable" (at no _extra_ cost).

> > ```C++
> > void
> > ConvertedCode(Yaml::Mapping &yaml)
> > {
> > // We could also use an explicit Yaml::kv() call here, but we do not have to
> > // because Yaml::Mapping knows that the first node it receives is the key
> > // and second node it receives is the value for that key (and so on).
> > yaml << "requests received" << stats.requests;
> > ```
 
> Is Yaml::Mapping derived from std::ostream?

Undecided, but most likely not.

> It must be, otherwise we could not rely on all the operator<<'s.

* I am not sure we want to rely on arbitrary operator "<<" to produce valid YAML snippets.
* AFAICT, if needed, we can forward all Yaml::Mapping::operator "<<" calls to std::ostream::operator "<<" calls with a few source code lines. It is easy to declare the corresponding wrapper template.

> How can we distinguish the first call from the second, third, and so on?  The number of calls per entry is not predetermined. 

Undecided, but we can, _for example_, increment a counter for each call or switch between two different stream types for odd/even calls. This feels like an implementation question rather than an API question. Let's come up with the best API we can think of and then tackle implementation difficulties -- I bet we would save time that way.

> We might have an Yaml::Mapping::operator() returning an ostream

I doubt switching to raw ostream is a good idea for general code. Raw stream access might be available for some exceptional cases, but we are currently not sketching out anything exceptional that needs such raw ostream access.

> > ```C++
> > const auto peers = yaml.openSequence("peers");
> > // TODO: We can automate this closure. Should we?
> > yaml.close(peers) // or peers.close();
> > ```
 
> We should. Why would we give up on RAII safety for the whole design to avoid opening one more c++ scope?

AFAICT, RAII safety cannot be reliably provided in this context: You may remember to always open and close C++ scope. I will forget. The compiler will not catch my mistake. Another developer will reuse my object later. A review may miss that bug. For example:

```C++
// use RAII for safety! (XXX: not really)
{
    auto peers = yaml.openSequence("peers"); // assume this returns an auto-closing sequence
    auto clients = yaml.openSequence("clients"); // assume this returns an auto-closing sequence
    // game over, but just for completeness sake:
    for (...) reportPeer(peers.entry(), peer);
    for (...) reportClient(clients.entry(), client);
    // RAII cleanup happens here, but that cleanup cannot fix broken code
}
```


> This example is not fitting, because it specializes Sequence, forcing it to Sequence-of-maps.

You lost me here. There is no sample code that _forces_ a Sequence to be a Sequence-of-maps AFAICT. While most sequences are sequences of maps, I suspect there are valid use cases where a sequence is a sequence of non-map nodes (I cannot immediately think of any though).


> It could maybe evolve into:
> 
> ```c++
>     const auto peers = yaml.openSequence("peers");
>     for (const auto &peer: CurrentPeers()) {
>        auto peerEntry = peers.entry().openMap();
>        peerEntry << "name" << peer->name;
>         if (peer->digest)
>             peer->digest->report(peers);
>     }
> ```

Sure, with one correction: The call to report() should get a `peerEntry` parameter, not `peers` parameter. I made the same mistake in my code sample. A compiler would have caught it, so this is not a big deal.

> .. It's starting to feel overcomplicated.

The above code looks rather simple/straightforward to me. In fact, it can be massaged to look almost exactly as LegacyCodeToBeConverted() loop if we assume that no-digest peers do not need to be reported at all. In retrospect, I should have made that assumption when sketching converted code.
 
> What if we just used [yaml-cpp](https://github.com/jbeder/yaml-cpp)?

AFAICT, that API is meant to support producing a single YAML scalar using multiple ostream calls but is dangerously inconsistent in its approach:

```C++
// very similar value-generating code produces drastically different results:
yaml << BeginMap << ... << Value << valuePrefix << valueSuffix; // OK: two "<<" calls for one value scalar
yaml << BeginSeq << valuePrefix << valueSuffix; // XXX: two "<<" calls produce two value scalars!
```

I doubt that such unflagged/raw/default scalar flexibility is a good idea for us because it is likely to lead to invalid YAML output. However, if manipulator-based API is indeed better, then we should either use that library or change our API accordingly. I speculate that we should do the latter in that hypothetical/unproven case.

>  Maybe it could be as simple as wrapping two of their manipulators to offer RAII safety

If you are talking about API _implementation_ details, then let's not discuss that until we settle on API.

Reply:
@kinkie, we have discussed how to resume YAML generation work during the conference call. AFAICT, the representative [test case](https://github.com/kinkie/indent/pull/1/files#r1664348635) (with both legacy code and future YAML output!) that we have built and discussed on this thread and the corresponding ConvertedCode() sketch are the right (re)starting point (_after_ applying a couple of fixes from your followup [review](https://github.com/kinkie/indent/pull/1/files#r1677187011) that I acknowledged in my [response](https://github.com/kinkie/indent/pull/1/files#r1677973728) to your review).